### PR TITLE
Implement ts.OrganizeImports.removeUnusedImports

### PR DESCRIPTION
**Pull request by amcasey**
TODO: Still need to add support for organizing imports in ambient
modules

### CODE REVIEW

File: src/services/organizeImports.ts [hunk 1]
```diff
@@ -47,8 +55,73 @@ namespace ts.OrganizeImports {
...
+            if (symbol === undefined) {
+                return true;
+            }
+
+            // The JSX factory symbol is always used.
+            if (jsxContext && symbol.name === jsxNamespace) {
+                return true;
+            }
+
+            const entries = FindAllReferences.getReferenceEntriesForNode(identifier.pos, identifier, program, [sourceFile], {
```
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: I can just add `.filter(e => e.type === "node" && e.node.getSourceFile() === sourceFile)`, but first I'd like to confirm that this isn't a bug in FAR.

**Comment by amcasey** [comment]
Failures are, once again, due to uppercase/lowercase mismatch.  Investigating.

File: src/services/organizeImports.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: It's coming in as a related symbol.   @mhegazy @andy-ms Would it be preferable to filter at the call site (as above) or add a property to `ts.FindAllReferences.Options`?

File: src/services/organizeImports.ts [hunk 2]
```diff
@@ -47,8 +55,73 @@ namespace ts.OrganizeImports {
...
+            }
+
+            if (name || namedBindings) {
+                usedImports.push(updateImportDeclarationAndClause(importDecl, name, namedBindings));
+            }
+        }
+
+        return usedImports;
+
+        function isDeclarationUsed(identifier: Identifier) {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: `return !!program.getTypeChecker().getSymbolAtLocation(identifier)!.isReferenced;`, and at top, `program.getSemanticDiagnostics(sourceFile); // Ensure we've checked this file so `isReferenced` is defined on symbols.`. Also must change check `isUse && result && nameNotFoundMessage && noUnusedIdentifiers` in `checker.ts` to not include `&& noUnusedIdentifiers`, so we always set `isReferenced` accurately.

File: src/services/organizeImports.ts [hunk 3]
```diff
@@ -47,8 +55,73 @@ namespace ts.OrganizeImports {
...
-    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>) {
-        return oldImports; // TODO (https://github.com/Microsoft/TypeScript/issues/10020)
+    function removeUnusedImports(oldImports: ReadonlyArray<ImportDeclaration>, sourceFile: SourceFile, program: Program) {
+        const typeChecker = program.getTypeChecker();
+        const jsxNamespace = typeChecker.getJsxNamespace();
+        const jsxContext = sourceFile.languageVariant === LanguageVariant.JSX && program.getCompilerOptions().jsx;
+
+        const usedImports: ImportDeclaration[] = [];
+
+        for (const importDecl of oldImports) {
```
Comment (andy-ms) [inline src/services/organizeImports.ts]: `mapDefined(usedImports, importDecl => <<<importDecl, updated import, or undefined>>>)`

File: src/services/organizeImports.ts (hunk 2 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: I'll defer to @mhegazy, who asked me to do it this way.

File: src/services/organizeImports.ts (hunk 3 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: I'm not sure I understand the advantage of doing it that way.

File: src/services/organizeImports.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amcasey) [reply src/services/organizeImports.ts]: Per our offline discussion, I've added filtering at the call site.

File: src/services/organizeImports.ts (hunk 2 above)
Comment (amcasey) [reply src/services/organizeImports.ts]: Any progress towards reaching a consensus on how to identify unused imports?

**Pull request merged**